prometheus_keys.lua

```diff
@@ -0,0 +1,128 @@
+-- Storage to keep track of used keys. Allows to atomically create, delete
+-- and list keys. The keys are synchronized between nginx workers
+-- using ngx.shared.dict. The whole purpose of this module is to avoid
+-- using ngx.shared.dict:get_keys (see https://github.com/openresty/lua-nginx-module#ngxshareddictget_keys),
+-- which blocks all workers and therefore it shouldn't be used with large
+-- amounts of keys.
+
+local KeyIndex = {}
+KeyIndex.__index = KeyIndex
+
+local lock_lib = require("prometheus_lock")
+
+function KeyIndex.new(shared_dict, prefix)
+  local self = setmetatable({}, KeyIndex)
+  self.dict = shared_dict
+  self.key_prefix = prefix .. "key_"
+  self.sync_hint_prefix = prefix .. "sync_"
+  self.key_count = prefix .. "key_count"
+  self.last = 0
+  self.keys = {}
+  self.index = {}
+  self.lock = lock_lib.new(prefix .. "lock_keys", self.dict)
+  return self
+end
+
+-- Loads new keys that might have been added by other workers since last sync.
+function KeyIndex:sync()
+  local full_sync_hint = self.dict:get(self.sync_hint_prefix .. ngx.worker.id()) or 0
+  local N = self.dict:get(self.key_count) or 0
+  if full_sync_hint > 0 then
```

Comment:
There is no code currently that decreasess `sync_hint_prefix`, so it seems that after the first key is deleted we'll be doing a full sync every time. Is that intentional?

Also, instead of keeping a separate sync hint counter per worker, I would suggest:
- a single monotonically increasing counter in the shared dictionary. After any deletion, the worker just increments this counter.
- a local variable storing the last sync hint value a worker previusly observed.

Something like this in `sync()`:

```lua
local sync_hint = self.dict:get(self.sync_hint_key) or 0
local N = self.dict:get(self.key_count) or 0
if sync_hint > self.sync_hint then
  self:sync_range(0, N)
  self.sync_hint = sync_hint
elseif N ~= self.last then
  self:sync_range(self.last, N)
end
return N
```

And in `remove_by_index`:
```lua
self.dict:incr(self.sync_hint_key, 1, 0)
self.sync_hint = self.sync_hint + 1
```

Reply:
You're right, I forgot to add the decrement. But your solution is actually much better, I'll implement that. Thanks!

Reply:
Done in dc0c60f5dea71de1aaaf5addb6f5d6398d1afad1.